testenv/utils/ssh.go

```diff
@@ -0,0 +1,171 @@
+		return nil, fmt.Errorf("failed to create private key file: %w", err)
+	}
+
+	privateKeyPEM := &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	}
+	if err := pem.Encode(privateKeyFile, privateKeyPEM); err != nil {
+		return nil, fmt.Errorf("failed to encode private key: %w", err)
+	}
+
+	if err := os.Chmod(privateKeyFile.Name(), 0o600); err != nil { //nolint:gomnd
+		return nil, fmt.Errorf("failed to change file permissions: %w", err)
+	}
+
+	publicKey, err := ssh.NewPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create public key: %w", err)
+	}
+
+	publicKeyFile := filepath.Join(workDir, "id_rsa_testenv.pub")
+	err = os.WriteFile(publicKeyFile, ssh.MarshalAuthorizedKey(publicKey), 0o600) //nolint:gomnd
+	if err != nil {
+		return nil, fmt.Errorf("failed to write public key file: %w", err)
+	}
+
+	return &SSHKeyPair{
+		PublicKeyFile:  publicKeyFile,
+		PrivateKeyFile: privateKeyFile.Name(),
+		Temporary:      true,
+	}, nil
+}
+
+// Run SSH tunnel to remote VMClarity server.
+func RunSSHTunnel(ctx context.Context, privateKeyFile, remoteHost, remotePort, localPort string) {
+	logger := GetLoggerFromContextOrDiscard(ctx)
+
+	// Read the private key file.
+	key, err := os.ReadFile(privateKeyFile)
+	if err != nil {
+		logger.Fatalf("failed to read private key: %s\n", err)
+		return
+	}
+
+	// Create Signer from private key.
+	signer, err := ssh.ParsePrivateKey(key)
+	if err != nil {
+		logger.Fatalf("failed to parse private key: %s\n", err)
+		return
+	}
+
+	// Create SSH client config.
+	config := ssh.ClientConfig{
+		User: "ubuntu",
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+		// TODO(paralta) Review this insecure host key callback, which accepts any host key.
+		HostKeyCallback:   ssh.InsecureIgnoreHostKey(), //nolint:gosec
+		HostKeyAlgorithms: []string{ssh.KeyAlgoED25519},
+	}
+
+	// Add port to remote host.
+	remoteAddress := fmt.Sprintf("%s:%s", remoteHost, "22")
+
+	// Dial the remote server.
+	client, err := ssh.Dial("tcp", remoteAddress, &config)
+	if err != nil {
+		logger.Fatalf("failed to dial: %s\n", err)
+	}
+	defer client.Close()
+
+	// Listen on local port.
+	listener, err := net.Listen("tcp", "localhost:"+localPort)
+	if err != nil {
+		logger.Fatalf("failed to listen: %s\n", err)
+	}
+	defer listener.Close()
+
+	for {
+		// Accept local connection.
+		local, err := listener.Accept()
+		if err != nil {
+			logger.Fatalf("failed to accept: %s\n", err)
+		}
+
+		// Dial remote server.
+		remote, err := client.Dial("tcp", "localhost:"+remotePort)
+		if err != nil {
+			logger.Fatalf("failed to dial: %s\n", err)
+		}
+
+		// Run tunnel between local and remote connections.
+		runTunnel(ctx, local, remote)
+	}
+}
+
+// runTunnel runs a tunnel between two connections.
+func runTunnel(ctx context.Context, local, remote net.Conn) {
+	logger := GetLoggerFromContextOrDiscard(ctx)
+
+	defer local.Close()
+	defer remote.Close()
```

Comment:
```suggestion
```

Reply:
If some reason the `runTunnel` is not called the connections will be leaked.

Reply:
addressed below in review